### DISCUSSION TITLE
List of Community Made Regex

### DISCUSSION THREAD

**Discussion by mattcoleanderson** [mattcoleanderson/obsidian-conceal-plugin#19] [Ideas]
## Purpose
Use this Discussion to share Regex you've made for concealing properties and text in your vault.

I hope to see this page fill up with helpful resources for many different plugins and use cases!

**Comment by BananenRitter** [comment]
# Conceal Example
The regex handles the following syntaxes:
- Single curly {brace}
- Double curly {{brace}}
- Multiple {matches} per {{line}}
- {c3|Prefixing} {c2::clozure} {{c1:number}}
- Doesnt ignore {{matches} with 2 braces on one side and a 1 on the other side
- Doesnt ignore {matches}} with 2 braces on one side and a 1 on the other side
- Ignores `{{matches}` within inline code blocks
```typescript
console. tog( 'Ignores {cl:matches} within {c2:multitiney codebtocks')
```


# Tipps: *(mostly useful for Regex Find and Replace Plugin)*
1. **Only combine a few regex commands at the same time** to make your Regex easily readable to prevent hours of searching for the mistake.
2. **Copy parts of your text below eachother to compare their structure** in order to find the right Syntax.
3. **Win+V/ Activate Clipboard** to keep more Regex Examples in Cashe while working with it.
4. **Save** your successfully made Regex Examples, you will most likely need them again.
5. **Power Toys** for Windows **Rename** can bulk rename with Regex & has many more usages. https://learn.microsoft.com/de-de/windows/powertoys/powerrename

# RegEx Find & Replace
## Updated 01.05.2024

Regular Expression
Kann *Links & LaTeX zerschießen* Gibt es einen Weg Links & LaTeX zu exkludieren?   (german)
- 

## Literature
https://en.wikipedia.org/wiki/Regular_expression

## Links  Helper Programs
- https://regex-generator.olafneumann.org/?sampleText=2020-03-12T13%3A34%3A56.123Z+INFO++%5Borg.example.Class%5D%3A+This+is+a+%23simple+%23logline+containing+a+%27value%27.&flags=i
- https://regex101.com/r/JqqII4/1
- https://regexr.com/
- https://www.regextester.com/

## Cheat sheet
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Cheatsheet


# Characters

| Regex basics                         | Description                                                                                          |
| ------------------------------------ | ---------------------------------------------------------------------------------------------------- |
| Empty Linke Break                    | ^\s*$                                                                                                |
| ^                                    | The start of a string                                                                                |
| $                                    | The end of a string                                                                                  |
| .                                    | Wildcard which matches any character, except newline (\n).                                           |
| \|                                   | Matches a specific character or group of characters on either side (e.g. a\|b corresponds to a or b) |
| \|Used to escape a special character |                                                                                                      |
| a                                    | The character "a"                                                                                    |
| ab                                   | The string "ab"                                                                                      |
| *Quantifiers*                        |                                                                                                      |
| *                                    | Used to match 0 or more of the previous (e.g. xy*z could correspond to "xz", "xyz", "xyyz", etc.)    |
| ?                                    | Matches 0 or 1 of the previous                                                                       |
| +                                    | Matches 1 or more of the previous                                                                    |
| {5}                                  | Matches exactly 5                                                                                    |
| {5,}                                 | Matches 5 or more occurrences of the preceding character or group                                    |
| {5, 10}                              | Matches everything between 5-10                                                                      |
| *Character classes*                  |                                                                                                      |
| \s                                   | Matches a whitespace character                                                                       |
| \S                                   | Matches a non-whitespace character                                                                   |
| \w                                   | Matches a word character                                                                             |
| \W                                   | Matches a non-word character                                                                         |
| \d                                   | Matches one digit                                                                                    |
| \D                                   | Matches one non-digit                                                                                |
| [\b]                                 | A backspace character                                                                                |
| \c                                   | A control character                                                                                  |
| *Special characters*                 |                                                                                                      |
| \n                                   | Matches a newline                                                                                    |
| \t                                   | Matches a tab                                                                                        |
| \r                                   | Matches a carriage return                                                                            |
| \ZZZ                                 | Matches octal character ZZZ                                                                          |
| \xZZ                                 | Matches hex character ZZ                                                                             |
| \0                                   | A null character                                                                                     |
| \v                                   | A vertical tab                                                                                       |
| *Groups*                             |                                                                                                      |
| (xyz)                                | Grouping of characters                                                                               |
| (?:xyz)                              | Non-capturing group of characters                                                                    |
| [xyz]                                | Matches a range of characters (e.g. x or y or z)                                                     |
| [^xyz]                               | Matches a character other than x or y or z                                                           |
| [a-q]                                | Matches a character from within a specified range                                                    |
| [0-7]                                | Matches a digit from within a specified range                                                        |
| *String replacements*                |                                                                                                      |
| $`                                   | Insert before matched string                                                                         |
| $'                                   | Insert after matched string                                                                          |
| $+                                   | Insert last matched                                                                                  |
| $&                                   | Insert entire match                                                                                  |
| $n                                   | Insert nth captured group                                                                            |
| *Assertions*                         |                                                                                                      |
| (?=xyz)                              | Positive lookahead                                                                                   |
| (?!xyz)                              | Negative lookahead                                                                                   |
| ?!= or ?<!                           | Negative lookbehind                                                                                  |
| \b                                   | Word Boundary (usually a position between /w and /W)                                                 |
| ?#                                   | Comment                                                                                              |

# Syntax
*von ChatGPT*
Explaining all regex syntax in one response might be quite extensive, but I can provide a brief overview of the most common elements used in regular expressions:
## Literals
are characters that match themselves. For example, the regex *a* matches the literal character "a".
start
## Character Classes `[]`
Character classes allow specifying *a set* of characters to match. Inside a character class, most characters lose their special meaning and match themselves literally. For example, the character class `[aeiou]` matches any vowel.
### 1. Ranges
You can specify a range of characters inside a character class using a hyphen *-*. For example, `[a-z]` matches any lowercase letter from 'a' to 'z'. Similarly, `[0-9]` matches any digit.
### 2. Negation
You can negate a character class by including a caret *^* as the first character inside the brackets. For example, `[^0-9]` matches any character that is not a digit.
### 3. Shorthand Notations
There are shorthand notations for common character classes:
- `\`*d*: Matches any *digit*,   equivalent to `[0-9]`
- `\`*D*: Matches any character that is *not digit*,   equivalent to `[0-9]`

- `\`*w*: Matches any *word character*,   equivalent to `[a-zA-Z0-9_]`
- `\`*W*: Matches any character that is *not word character*,   equivalent to `[a-zA-Z0-9_]`

- `\`*s*: Matches any *whitespace character* (spaces, tabs, newlines, etc.).
- `\`*S*: Matches any character that is *not whitespace character*.

- `\`*b*: Matches a *word boundary* (the position between a word character and a non-word character).
- `\`*B*: Matches a position that is *not word boundary*.
### 4. Escaping Inside Character Classes
Inside a character class, certain characters retain their special meanings and need to be escaped to match them literally. For example, to match a literal hyphen *-*, you would escape it like this: *`\`-*.

## Quantifiers
specify how many times a character or group of characters can occur. Some common quantifiers are:
### `*`   0 or More     Asterisks
match the preceding element **zero or more times**. For example, *a* matches "", "a", "aa", "aaa", and so on.
### `+`   1 or More
Matches the preceding element **one or more times**. For example, *a+* matches "a", "aa", "aaa", and so on, but not "".
### `?`   Optional
Matches the preceding element **zero or one time. It makes the preceding element optional.** For example, *a?* matches "", "a", but not "aa" or any longer sequence.
#### `+?` Match Non Greedy
- In regex, quantifiers like *+* indicate repetition. Specifically, *+* means "one or more times". When used with capturing parentheses *( )*, as in *([`\`w`\`d`\`s`\`S]+?)*, it means that the pattern inside the parentheses should be matched one or more times, but as few times as possible, i.e., in a non-greedy or lazy way.
*let's consider the pattern* `[`\`w`\`d`\`s`\`S]+? *applied to the string* "abc123!@#$".
- Without the *?* (greedy): `[`\`w`\`d`\`s`\`S]+* would match the entire string "abc123!@#$".
- With the *?* (non-greedy): `[`\`w`\`d`\`s`\`S]+?* would match "a" first, then "b", then "c", and so on, until it reaches the end of the string.

### `{n}`    Exactly
Matches **exactly** *n* occurrences of the preceding element.
For example, *a{3}* matches "aaa" but not "aa" or any longer sequence.
### `{n,}`  At Least
Matches **at least** *n* occurrences of the preceding element.
For example, *a{2,}* matches "aa", "aaa", "aaaa", and so on.
### `{n,m}` Between
Matches **between** *n* and *m* occurrences of the preceding element, inclusive.
For example, *a{2,4}* matches "aa", "aaa", or "aaaa", but not "a" or "aaaaa".

## Anchors
are used to specify positions in the text.
### `^`  Start Line   caret
anchors the regex at the **start of a line.** It matches the beginning of the string or the start of a line when using multi-line mode.
### `$`  End Line
anchors the regex at the **end of a line.** It matches the end of the string or the end of a line when using multi-line mode.
### `\b` Asymmetrical   Word Boundary
matches a **position where one side is a word character** (such as a letter, digit, or underscore), **and the other side is not a word character** (such as whitespace or punctuation). It doesn't consume any characters, only matches a position.
### `\B` Symmetrical   Non-Word Boundary
matches a **position where both sides are word characters** or **both sides are non-word characters.** It also doesn't consume any characters, only matches a position.

## Alternation `|` 
- is represented by the *|* symbol and allows for matching either of two alternatives. For example, *cat|dog* matches either "cat" or "dog".
- Alternation is a powerful feature of regular expressions that allows you to specify multiple options for matching text. It's commonly used in search and replace operations and data validation tasks.

### Order of Evaluation
Alternation evaluates from **left to right**, and the first alternative that matches is selected.
For example, in the pattern *foo|foobar*, if the text contains "foobar", it will match "foo" because it's the first alternative.
### Grouping Alternatives
Parentheses *( )* can be used to group alternatives together. This is useful when applying quantifiers or other operators to multiple alternatives. For example, *(cat|dog)+* matches one or more occurrences of "cat" or "dog".
### Backtracking
When using alternation, the regex engine backtracks to try all possible combinations of alternatives until a match is found. This can impact performance if there are many alternatives or if the text being searched is large.

## Parentheses `()`
Parentheses are used to **group characters or expressions together.** This allows applying quantifiers or alternation to a group. For instance, *(abc)+* matches "abc", "abcabc", "abcabcabc", and so forth.
### Capturing Groups:
Parentheses create capturing groups, which capture the matched text for later use.
In the regex *(foo)bar, (foo)* captures "foo", and *(foo)bar* matches "foobar", capturing "foo".
### `?:`Non-capturing Groups:
Using *(?: )*, you can create non-capturing groups. These function like regular groups but don't capture matched text. Useful for applying quantifiers or alternation without capturing text.
For example, *(?:foo)bar* matches "foobar" without capturing "foo".
### Backreferences:
Capturing groups can be referenced later in the regex or in replacement patterns using backreferences.
For instance, *`\`*1 refers to the first capturing group, *`\`*2 to the second, and so forth.*
#### Example
Regex: *(foo(bar))*
- Matches: "foobar"
- Capturing groups:
    - Group 1: "foobar"
    - Group 2: "bar"
*Group 2 is a capturing group within another capturing group (Group 1), demonstrating nesting of capturing groups.*

## Escape Sequences & Character Escapes `\`
- Some characters have special meanings in regex and need to be escaped to match them literally. For example, *`\`.* matches a literal period.
- Escape sequences are essential for matching special characters literally and for handling metacharacters in regular expressions.

### Matching Special Characters Literally
Some characters have special meanings in regular expressions. To match them literally, you precede them with a backslash `\`.
### Examples of Escape Sequences
- *`\`.*: Matches a literal period *.*.
- `\*`: Matches a literal asterisk `*`.
- *`\`^*: Matches a literal caret *^*.
- *`\$`*: Matches a literal dollar sign.
### `\` , `\\` Escaping Meta-Characters
Escape sequences are also used to escape metacharacters, which have special meanings in regex.
For example, `\`*d* matches any digit,
but `\\`*d* matches a literal backslash followed by the letter "d".
### `[]`Escaping Inside Character Classes
Inside a character class `[]`, some characters retain their special meanings and need to be escaped to match them literally. For example, to match a literal hyphen *-*, you would escape it like this: *`\`-*.

## Modifiers
Modifiers in regular expressions are **flags that change the behavior of the pattern matching process.** They affect how the pattern is applied to the text being searched.

### `(i)` Case Insensitivity
The *i* modifier makes the pattern match case-insensitively.
For example, */hello/i* matches "hello", "Hello", "HELLO", etc.
### `(g)` Global Matching
The *g* modifier performs a global search, meaning it **finds all matches rather than stopping after the first match.**
For example, */a/g* matches all occurrences of "a" in the text.
### `(m)` Multiline Mode
The *m* modifier **changes the behavior** of anchors *^* and *$* to match the start and end of lines within a multiline string.
For example, */^hello/m* matches "hello" only at the start of a line.
### `(s)` Dotall Mode
The *s* modifier makes the dot *.* **match all characters, including newline characters** *`\`n*. By default, the dot matches any character except newline.
For example, */hello.world/s* matches "hello`\`nworld".
### `(u)` Unicode Mode
The *u* modifier **enables full Unicode matching.** It allows regex patterns to work with Unicode characters and properties. For example, *`\`p{Letter}* matches any Unicode letter.

## Assertions
- are used to assert whether a **pattern is found** at a particular position in the text **without consuming any characters.** 
- Assertions are useful for specifying conditions on the text without actually consuming any characters. They allow you to match text based on its context or surroundings.

### `(?=...)` Positive Lookahead
Asserts that a certain pattern exists after the current position in the text, without including it in the match.
For example, *foo(?=bar)* matches "foo" **only if it is followed by** "bar".
### `(?!...)` Negative Lookahead
Asserts that a certain pattern does not exist after the current position in the text.
For example, *foo(?!bar)* matches "foo" **only if it is not followed by** "bar".
### `(?<=...)` Positive Lookbehind
Asserts that a certain pattern exists before the current position in the text, without including it in the match.
For example, *(?<=foo)bar* matches "bar" **only if it is preceded by** "foo".
### `(?<!...)` Negative Lookbehind
Asserts that a certain pattern does not exist before the current position in the text.
For example, *(?<!foo)bar* matches "bar" **only if it is not preceded by** "foo".
# Regex Examples
## 1   `\n\n\{\n[a-z]{#}"`
### Explanation:
- `\n\n` matches two consecutive line breaks.
- `\{` matches the opening curly brace `{`.
- `\n` matches a line break.
- `[a-z]` matches any lowercase letter.
- `{#}` is a placeholder for the number of characters you want to match. You should replace `#` with the actual number you want to match.
- `"` matches the closing double quote.

Make sure to replace `#` with the specific number of characters you expect to find between the lowercase letters and the closing double quote.

## 2   `{\n[a-z]#\}"`
### Explanation:
- An opening curly brace `{`.
- A lowercase letter.
- The hashtag symbol `#`.
- A closing double quote `"`.

## 3   `{\n[a-z][!@#$%^&*()]"`
To match arbitrary special characters in regex, you can use character classes or escape them with a backslash. Here's how you can do it:
1. **Character Classes**: You can place special characters inside square brackets `[]`. Inside square brackets, most characters lose their special meaning and are treated literally.
6. **Escaping with Backslash**: You can also escape special characters with a backslash `\`. This tells the regex engine to treat the character literally.
For example, to match arbitrary special characters, you could use a pattern like this:
### Explanation
- An opening curly brace `{`.
- A lowercase letter.
- Any special character from the set `!@#$%^&*()`.
- A closing double quote `"`.

If you want to match a specific special character, you escape it with a backslash. For example, to match the hashtag symbol `#`, you would use `\#`:

## 4   `{\n[a-z]\#"` 
### Explanation
- An opening curly brace `{`.
- A lowercase letter.
- The hashtag symbol `#`.
- A closing double quote `"`.

## 5   `\{[\w\d\s\S]+?\n`
### Explanation:
3. `\{`: Matches the opening curly brace `{`.
7. `[\w\d\s\S]+?`: Matches any character (word characters, digits, whitespace, and non-whitespace characters) one or more times in a non-greedy way.
8. `\n`: Matches another line break to indicate the end of the match.

This pattern will match each block of text enclosed within curly braces, spanning from one line break after the opening curly brace to the next line break.
## 6 Find & Capture`(\n([a-zA-Z]+))`
### Explanation:
- `(\n([a-zA-Z]+))`: This pattern captures a newline `\n` followed by at least one letter (uppercase or lowercase) inside a capturing group.
### Replace with Captured`## $2 \n`
- `## ` : Inserts "## " before the captured letters.
- `$2`: Inserts the content of the second capturing group, which represents the letters captured in the newline.
Put together, this search and replace operation will add "## " before any newline that contains at least one letter.
For example, in Python, you can use this regex pattern with the `re.sub()` function to perform the replacement:
- `\n`: Inserts a linebreak after the captured letters.
## 7 `\b(\w+)\b(?:\n\1\b)+`
### Replace: 
`$1`

### Explanation:
- `\b`: Word boundary.
- `(\w+)`: Matches and captures a word.
- `\b`: Word boundary.
- `(?:\n\1\b)+`: Non-capturing group that matches a newline character followed by the previously captured word and a word boundary, repeated one or more times.

# Java Regex
- Java characters that have to be escaped in regular expressions are:
    `\.[]{}()<>*+-=!?^$|`
- Two of the closing brackets (`]` and `}`) only have to be escaped after opening the same type of bracket.
- In `[]`-brackets some characters (like `+` and `-`) do sometimes work without escape.
- http://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html

**Comment by BananenRitter** [comment]
Example from Dev as Text

# Conceal Example
The regex handles the following syntaxes:
- Single curly {brace}
- Double curly {{brace}}
- Multiple {matches} per {{line}}
- {c3|Prefixing} {c2::clozure} {{c1:number}}
- Ignores `{{matches}` within inline code blocks
```typescript
console. tog( 'Ignores {cl:matches} within {c2:multitiney codebtocks')
